fix: onnx PReLU unidirectional broadcasting

ganler opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/70570

ganler committed:
fix: onnx PReLU unidirectional broadcasting

thiagocrepaldi reviewed (commented):
LGTM, could you also add a unit test for this case?

ganler committed:
add assertion to ensure unidirectional broadcasting in PReLU

ganler committed:
add test case for prelu with scalar tensor

ganler commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -719,6 +719,9 @@ def prelu(g, self, weight):
         weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+    elif self_rank is not None and self_rank == 0:
+        self = sym_help._unsqueeze_helper(g, self, [0])
+    assert sym_help._get_tensor_rank(self) >= len(sym_help._get_tensor_sizes(weight)), "slope must be unidirectionally broadcastable to input tensor"
```
The bug needs to be reproduced using TVM but will introduce a huge testing dependency. So instead I added a rank checking for it.

ganler committed:
check none

ganler commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -719,6 +719,9 @@ def prelu(g, self, weight):
         weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+    elif self_rank is not None and self_rank == 0:
+        self = sym_help._unsqueeze_helper(g, self, [0])
+    assert self_rank is None or sym_help._get_tensor_rank(self) >= len(sym_help._get_tensor_sizes(weight)), "slope must be unidirectionally broadcastable to input tensor"
```
The bug needs to be reproduced using TVM but will introduce a huge testing dependency. So instead I added a rank checking for it.

ganler committed:
flake8

ganler committed:
fix unittest

check none

fix indent

fix none (cont)

fix python err

ganler committed:
trigger ci

ganler commented:
timeout tests seem to be flaky. pushed an empty commit to trigger the ci.

ganler committed:
Merge branch 'master' into onnx-prelu

ganler commented:
Hi @thiagocrepaldi , could you also double-check the added test case? Thank you!

garymm commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -6034,6 +6034,12 @@ def forward(self, x):
                       dynamic_axes={"x": [1, 2]},
                       test_with_inputs=[y])
 
+    def test_prelu_scalar(self):
```
Does this test fail before this change and pass afterwards? If not then I don't think this test is useful.

garymm reviewed (commented):
I don't think this is actually needed. See my comment here:
https://github.com/pytorch/pytorch/issues/70570#issuecomment-1026258821

ganler committed:
refact: resolve comments

ganler commented:
Hi @garymm , thanks for the review and I have addressed the comments. 

Regarding the benefit of this PR, this change will help:
1. TVM accepts the model in https://github.com/pytorch/pytorch/issues/70570#issue-1092220771
2. ORT reduces the warning of "Expected shape from model of {} does not match actual shape of {1} for output 2"

garymm commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -717,11 +717,19 @@ def squeeze(g, self, dim=None):
+            self = sym_help._unsqueeze_helper(g, self, [0])
+            self_rank = sym_help._get_tensor_rank(self)
+
+    slope_rank = sym_help._get_tensor_rank(weight)
```
Don't mix torch names and ONNX names, it's confusing.

```suggestion
    weight_rank = sym_help._get_tensor_rank(weight)
```

garymm commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -717,11 +717,19 @@ def squeeze(g, self, dim=None):
+            weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+        elif self_rank == 0:
+            self = sym_help._unsqueeze_helper(g, self, [0])
+            self_rank = sym_help._get_tensor_rank(self)
```
Don't we know this for sure?
```suggestion
            self_rank = 1
```

garymm commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -717,11 +717,19 @@ def squeeze(g, self, dim=None):
-        weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+    if self_rank is not None:
+        if self_rank > 2:
+            weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
```
```suggestion
            # make weight unidirectional broadcastable
            weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
```

garymm commented on torch/onnx/symbolic_opset9.py:
```diff
@@ -717,11 +717,19 @@ def squeeze(g, self, dim=None):
+        if self_rank > 2:
+            weight = sym_help._unsqueeze_helper(g, weight, list(range(1, self_rank - 1)))
+        elif self_rank == 0:
+            self = sym_help._unsqueeze_helper(g, self, [0])
```
```suggestion
            # weight is always rank 1. torch allows scalar self, and ONNX is ambiguous
            # about whether this is allowed, but some implementations enforce
            # rank(self) >= rank(weight), which makes sense.
            self = sym_help._unsqueeze_helper(g, self, [0])
```

garymm commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -6034,6 +6034,7 @@ def forward(self, x):
                       dynamic_axes={"x": [1, 2]},
                       test_with_inputs=[y])
```
Add a test that uses scalar inputs to trigger the new code path.
Even if the difference between the old behavior and new behavior is just "ORT doesn't print a warning", that's OK. Better to have a test that executes the code path and makes sure it basically works.

garymm reviewed (changes requested):
Thanks for the contribution, and sorry for the latency.

ganler committed:
Update torch/onnx/symbolic_opset9.py

Co-authored-by: Gary Miguel <garymm@garymm.org>

ganler committed:
Update torch/onnx/symbolic_opset9.py

Co-authored-by: Gary Miguel <garymm@garymm.org>

ganler committed:
Update torch/onnx/symbolic_opset9.py

Co-authored-by: Gary Miguel <garymm@garymm.org>

ganler committed:
rename weight (slope_rank in ONNX)

ganler committed:
unittest for prelu-scalar

garymm commented on test/onnx/test_pytorch_onnx_onnxruntime.py:
```diff
@@ -6034,6 +6034,11 @@ def forward(self, x):
+    def test_prelu_scalar(self):
+        x = torch.scalar_tensor(1.)
+        self.run_test(torch.nn.PReLU(), x, input_names=["x"],
+                      test_with_inputs=[x])
```
Remove test_with_inputs. By default `run_test` will test with the inputs in the second argument, so this is redundant.

ganler committed:
rm test_with_input argument in prelu-scalar test

garymm reviewed (approved):
Will merge once checks pass.

garymm commented:
@msaroufim could you please import? (non-MS contribution, can't use merge bot yet)

garymm commented:
@pytorchbot merge this
